%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Finishing and Executing the Workflow}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}
    \frametitle{Outline}
    \begin{columns}[t]
        \begin{column}{.5\textwidth}
            \tableofcontents[sections={1-9},currentsection]
        \end{column}
        \begin{column}{.5\textwidth}
            \tableofcontents[sections={10-18},currentsection]
        \end{column}
    \end{columns}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}
  \frametitle{What is this about?}
   \question[Questions]{So far our workflow will not run properly. Why? What do we need to change?}
   \docs[Objectives]{\begin{enumerate}
                      \item Understanding the target idea.
                      \item Running simple workflows.
                     \end{enumerate}}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Adding a Target Rule}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}
  \frametitle{Why Target Rules}
  So far, we always executed the workflow by specifying a target file at the command line. How cumbersome!\newline
  \docs{We remember: \texttt{Snakemake} will automatically determine for a given rule, which expected outcomes are missing and execute all necessary rules, accordingly.\newline\pause
        The ``trick'' is that a workflow can have a ``target'' rule, which specifies the \emph{final} output(s) of a workflow. Any invokation of \texttt{Snakemake} will then execute \emph{all} rules of a workflow.}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{Best Practice}
  \docs{If no target is given at the command line, \texttt{Snakemake} will define the first rule of the \texttt{Snakefile} as the target.}
  Conventially, this rule is named \altverb{all}. This means that we add a rule at the top of our workflow:\newline
  \begin{onlyenv}<1| handout:0>
   \question{Which is our target file?}
   \begin{lstlisting}[language=Python,style=Python]
 rule all:
    input: 
   \end{lstlisting}
  \end{onlyenv}
  \begin{onlyenv}<2| handout:1>
   Our target rule is:
   \begin{lstlisting}[language=Python,style=Python]
 rule all:
    input: 
        "plots/quals.svg"
   \end{lstlisting}
  \end{onlyenv}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection{Running the final Workflow}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{Our ``final'' Workflow}
  You will find the workflow at XXX
  \begin{columns}
    \begin{column}{0.5\textwidth}
      \begin{lstlisting}[language=Python,style=Python,basicstyle=\tiny]
SAMPLES = ["A", "B"]

rule all:
    input:
        "plots/quals.svg"


rule bwa_map:
    input:
        "data/genome.fa",
        "data/samples/{sample}.fastq"
    output:
        "mapped_reads/{sample}.bam"
    shell:
        "bwa mem {input} | samtools view -Sb"
        " - > {output}"

rule samtools_sort:
    input:
        "mapped_reads/{sample}.bam"
    output:
        "sorted_reads/{sample}.bam"
    shell:
        "samtools sort"
        " -T sorted_reads/{wildcards.sample}"
        " -O bam {input} > {output}"
      \end{lstlisting}
    \end{column}
    \begin{column}{0.5\textwidth}
      \begin{lstlisting}[language=Python,style=Python,basicstyle=\tiny]
rule samtools_index:
    input:
        "sorted_reads/{sample}.bam"
    output:
        "sorted_reads/{sample}.bam.bai"
    shell:
        "samtools index {input}"

rule bcftools_call:
    input:
        fa="data/genome.fa",
        bam=expand("sorted_reads/{sample}.bam",
            sample=SAMPLES),
        bai=expand("sorted_reads/{sample}.bam.bai",
            sample=SAMPLES)
    output:
        "calls/all.vcf"
    shell:
        "bcftools mpileup -f {input.fa}"
        " {input.bam} | "
        "bcftools call -mv - > {output}"

rule plot_quals:
    input:
        "calls/all.vcf"
    output:
        "plots/quals.svg"
    script:
        "scripts/plot-quals.py"
      \end{lstlisting}
    \end{column}
  \end{columns}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{\HandsOn{Executing this Workflow}}
  Does our Workflow contain errors? We run a debug trial:
  \begin{lstlisting}[language=Bash, style=Shell]
$ snakemake --debug
  \end{lstlisting}
  \pause
  Some targets are already present, we want the entire workflow again:
  \begin{lstlisting}[language=Bash, style=Shell]
$ snakemake -j4 --forcerun
  \end{lstlisting}
  \question{What do you observe? Why \texttt{-j4}?}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
  \frametitle{\HandsOn{Visualising the Output}}
  On \mogon{} the linux distro is CentOS (or now AlmaLinux), which provides the \altverb{display}-program to display simple images. We shall invoke:
  \begin{lstlisting}[language=Bash, style=Shell]
$ display plots/quals.svg
  \end{lstlisting}
  The figure has no axis-labels.
  \question{What does the figure display?}
\end{frame}
